### PR TITLE
Add `search-in-current-path` feature

### LINKED ISSUE #3612: Button to search in current path
We can [search by file location](https://docs.github.com/en/free-pro-team@latest/github/searching-for-information-on-github/searching-code#search-by-file-location) using `path:foo`. It may be helpful to make search scoped to the current path by default, so we don't have to type the path by hand or copy and paste.

URL the feature should appear:

- `isRepoTree` (https://github.com/sindresorhus/refined-github/tree/master/source)
- `isSingleFile` (https://github.com/sindresorhus/refined-github/blob/master/source/background.ts)

Some thoughts:

1. If the user has modified the search, then it will be persisted
2. It should not activate on `isRepoRoot`
3. For files, search in its parent path (e.g. `path:foo` when viewing `foo/bar.ts`)
   - And according to 2, it should not activate on root level files (e.g. `/foo.ts`)

## Demo

1tqWPWcmQP

<details>
<summary>A rough implementation</summary>

```ts
import select from 'select-dom';
import onetime from 'onetime';
import * as pageDetect from 'github-url-detection';

import features from '.';
import {getCleanPathname} from '../github-helpers';

let search = '';

function getSearch(): string {
	const path = getCleanPathname().split('/');

	if (pageDetect.isRepoTree() && !pageDetect.isRepoRoot()) {
		return `path:${path.slice(4).join('/')} `;
	}

	if (pageDetect.isSingleFile() && path.length !== 5) {
		return `path:${path.slice(4, path.length - 1).join('/')} `;
	}

	return ''
}

function setSearch(): void {
	const searchInput = select<HTMLInputElement>('[data-hotkey="s,/"]')!;

	if (searchInput.value === search) {
		search = getSearch();
		searchInput.value = search;
	}
}

function init(): void {
	setSearch();
	document.addEventListener('pjax:end', setSearch);
}

features.add({
	id: __filebasename,
	description: '',
	screenshot: '',
}, {
	include: [
		pageDetect.isRepoTree,
		pageDetect.isSingleFile
	],
	init: onetime(init)
});
```

</details>

### PR DESCRIPTION
**Pull request by kidonng**
1. LINKED ISSUES:
   Resolve #3612

2. TEST URLS:
   It should activate on:

	- Repo tree: https://github.com/sindresorhus/refined-github/tree/master/source
	- Single file: https://github.com/sindresorhus/refined-github/blob/master/source/background.ts
	- Sync with current path: https://github.com/sindresorhus/refined-github/tree/master/source -> click a folder 
	
	It should not activate on:
	
	- Repo root: https://github.com/sindresorhus/refined-github
	- Root level files: https://github.com/sindresorhus/refined-github/blob/master/contributing.md
	- When the search box's content has changed

3. SCREENSHOT:
   https://user-images.githubusercontent.com/44045911/96068679-29f6ee80-0ecf-11eb-8c6b-af2e401dc5bf.gif

### CODE REVIEW

File: source/features/search-in-current-path.tsx [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+
+	if (searchInput.value === search) {
+		search = getSearch();
+		searchInput.value = search;
+	}
+}
+
+function init(): void {
+	setSearch();
+	document.addEventListener('pjax:end', setSearch);
```
Comment (fregante) [inline source/features/search-in-current-path.tsx:35]: We already handle this event. Avoid `onetime` and possibly use `repeatOnBackButton` instead.
`setSearch` should be the `init`

File: source/features/search-in-current-path.tsx [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+
+import features from '.';
+import {getCleanPathname} from '../github-helpers';
+
+let search = '';
+
+function getSearch(): string {
+	const path = getCleanPathname().split('/');
+
+	if (pageDetect.isRepoTree() && !pageDetect.isRepoRoot()) {
```
Comment (fregante) [inline source/features/search-in-current-path.tsx:13]: You can move `pageDetect.isRepoRoot` to `exclude`

File: source/features/search-in-current-path.tsx [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+		return `path:${path.slice(4).join('/')} `;
+	}
+
+	if (pageDetect.isSingleFile() && path.length !== 5) {
```
Comment (fregante) [inline source/features/search-in-current-path.tsx:17]: 5?

File: source/features/search-in-current-path.tsx [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+	return '';
+}
+
+function setSearch(): void {
+	const searchInput = select<HTMLInputElement>('[data-hotkey="s,/"]')!;
+
+	if (searchInput.value === search) {
+		search = getSearch();
+		searchInput.value = search;
+	}
```
Comment (fregante) [inline source/features/search-in-current-path.tsx:30]: Why do all this instead of just `searchInput.value = getSearch()`?

Comment (kidonng) [reply source/features/search-in-current-path.tsx:21]: > It should not activate on:
> * When the search box's content has changed

Comment (fregante) [reply source/features/search-in-current-path.tsx:21]: This should be in a JavaScript comment

Comment (fregante) [reply source/features/search-in-current-path.tsx:21]: But in what case would the content be changed *and then* the user changes the page?

File: source/features/search-in-current-path.tsx [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+function init(): void {
+	const searchInput = select<HTMLInputElement>('[data-hotkey="s,/"]')!;
+
+	if (searchInput.value === search) {
+		const {route, filePath} = new GitHubURL(location.href);
+		
+		if (route === 'tree') {
+			searchInput.value = `path:${filePath} `;
+		} else if (filePath.includes('/')) { // Exclude files at the root
+			searchInput.value = `path:${filePath.slice(0, filePath.lastIndexOf('/'))} `;
```
Comment (fregante) [inline source/features/search-in-current-path.tsx:18]: I haven't tested my last commit, but this should make it work correctly on files on the root of branches with slashes. `GitHubURL` was specifically built to consider this.

Test URL: https://github.com/sindresorhus/refined-github/blob/slash/dot/license

File: source/features/search-in-current-path.tsx [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+		}
+	}
+}
+
+void features.add({
+	id: __filebasename,
+	description: 'Adds current path to the search box.',
+	screenshot: 'https://user-images.githubusercontent.com/44045911/94982458-1cf00c00-056d-11eb-852a-4326042354b2.gif'
```
Comment (fregante) [inline source/features/search-in-current-path.tsx:26]: Can you make a gif without OctoTree enabled?

File: source/features/search-in-current-path.tsx [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+			searchInput.value = `path:${filePath} `;
+		} else if (filePath.includes('/')) { // Exclude files at the root
+			searchInput.value = `path:${filePath.slice(0, filePath.lastIndexOf('/'))} `;
+		}
+	}
+}
+
+void features.add({
+	id: __filebasename,
+	description: 'Adds current path to the search box.',
```
Comment (fregante) [inline source/features/search-in-current-path.tsx:25]: Needs a better description

File: source/features/search-in-current-path.tsx [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+		searchInput.value = search;
+	}
+}
+
+void features.add({
+	id: __filebasename,
+	description: 'Adds current path to the search box.',
+	screenshot: 'https://user-images.githubusercontent.com/44045911/94982458-1cf00c00-056d-11eb-852a-4326042354b2.gif'
+}, {
+	repeatOnBackButton: true,
```
Comment (yakov116) [inline source/features/search-in-current-path.tsx:30]: Nitpick can you put this on top of the init?

**Comment by kidonng** [comment]
😅 Sorry for the force push, I messed up the tree again

File: source/features/search-in-current-path.tsx (hunk 5 above)
Comment (kidonng) [reply source/features/search-in-current-path.tsx]: Tested, it works well 🎉

File: readme.md [hunk 9]
```diff
@@ -134,6 +134,7 @@ Thanks for contributing! 🦋🙌
 - [](# "sticky-sidebar") [Makes conversation sidebars and repository sidebars sticky, if they fit the viewport.](https://user-images.githubusercontent.com/10238474/62276723-5a2eaa80-b44d-11e9-810b-ff598d1c5c6a.gif)
 - [](# "link-to-github-io") [Add a link to visit the user’s github.io website from its repo.](https://user-images.githubusercontent.com/31387795/94045261-dbcd5e80-fdec-11ea-83fa-30bb673cc26e.jpg)
 - [](# "next-scheduled-github-action") [Shows the next scheduled time of relevant GitHub Actions in the workflows sidebar.](https://user-images.githubusercontent.com/46634000/94690232-2476a180-0330-11eb-99d7-e174bb762cea.png)
+- [](# "search-in-current-path") [Adds current path to the search box.](https://user-images.githubusercontent.com/44045911/94982458-1cf00c00-056d-11eb-852a-4326042354b2.gif)
```
Comment (yakov116) [inline readme.md:137]:
```suggestion
- [](# "search-in-current-path") [Adds current path to the search box.](https://user-images.githubusercontent.com/44045911/96068679-29f6ee80-0ecf-11eb-8c6b-af2e401dc5bf.gif)
```

Comment (fregante) [reply readme.md:137]: Looks good! I wonder if actually showing what that means would be beneficial. For example change folder, then search a file to show that the search is limited to that folder.

**Comment by fregante** [comment]
Feel free to send a new PR when you have time

**Comment by fregante** [comment]
ping @kidonng 😁

**Comment by kidonng** [comment]
Wait for it 😜 this is always on my mind

**Pull request closed without merging**